oauth2/cookies.go

```diff
@@ -0,0 +1,81 @@
+package oauth2
+
+import (
+	"context"
+	"net/http"
+	"time"
+)
+
+const (
+	// DefaultCookieName is the name of the stored cookie
+	DefaultCookieName = "session"
+)
+
+var _ Authenticator = &cookie{}
+
+// cookie represents the location and expiration time of new cookies.
+type cookie struct {
+	Name     string
+	Duration time.Duration
+	Now      func() time.Time
+	Tokens   Tokenizer
+}
+
+// NewCookieJWT creates an Authenticator that uses cookies for auth
+func NewCookieJWT(secret string, duration time.Duration) Authenticator {
+	return &cookie{
+		Name:     DefaultCookieName,
+		Duration: duration,
+		Now:      time.Now,
+		Tokens: &JWT{
+			Secret: secret,
+			Now:    time.Now,
+		},
+	}
+}
+
+// Validate returns Principal of the Cookie if the Token is valid.
+func (c *cookie) Validate(ctx context.Context, r *http.Request) (Principal, error) {
+	cookie, err := r.Cookie(c.Name)
+	if err != nil {
+		return Principal{}, ErrAuthentication
+	}
+	return c.Tokens.ValidPrincipal(ctx, Token(cookie.Value), c.Duration)
+}
+
+// Authorize will create cookies containing token information.  It'll create
+// a token with cookie.Duration of life to be stored as the cookie's value.
+func (c *cookie) Authorize(ctx context.Context, w http.ResponseWriter, p Principal) error {
+	token, err := c.Tokens.Create(ctx, p, c.Duration)
```

Comment:
So if I understand this, the cookie will be valid for the `c.Duration`, and the JWT `ExpiresAt` claim will also be set to `Now()`+`c.Duration` ? If so, I like this because it prevents clients from cheating and I think it reduces the feasibility of replays after the `ExpiresAt` claim time.